### PR TITLE
ENH Adds no_batch_dim tests/docs for LPPool1d and Identity

### LINKED ISSUE #60585: Rollup: No-batch-dim support for torch.nn modules
## Background
A previous issue (https://github.com/pytorch/pytorch/issues/47149) requested support of arbitrary batch dimensions across module inputs. This rollup issue details the `torch.nn` module updates required to address a subset of this functionality: specifically, the case of no batch dimensions. This particular case is useful for composability with a future vmap.

## Module support
- [x] `nn.AdaptiveAvgPool1d` - Should be updated to support 2D inputs
- [x] `nn.AdaptiveAvgPool2d` - Supports 3D inputs (no batch), but docs should make this clear
- [x] `nn.AdaptiveAvgPool3d` - Supports 4D inputs (no batch), but docs should make this clear
- [x] `nn.AdaptiveLogSoftmaxWithLoss` - Should be updated to support 1D input and scalar target / output1 (@george-qi)
- [x] `nn.AdaptiveMaxPool1d` - Should be updated to support 2D inputs
- [x] `nn.AdaptiveMaxPool2d` - Supports 3D inputs (no batch), but docs should make this clear
- [x] `nn.AdaptiveMaxPool3d` - Supports 4D inputs (no batch), but docs should make this clear
- [x] `nn.AlphaDropout`
- [x] `nn.AvgPool1d` - Should be updated to support 2D inputs
- [x] `nn.AvgPool2d` - Supports 3D inputs (no batch), but docs should make this clear
- [x] `nn.AvgPool3d` - Supports 4D inputs (no batch), but docs should make this clear
- [x] `nn.BCELoss`
- [x] `nn.BCEWithLogitsLoss`
- [x] `nn.Bilinear` - Already supports no batch dims, but docs should make this clear (@george-qi)
- [x] `nn.CELU`
- [x] `nn.CTCLoss` - Should be updated to support a single (source, target) sequence (@george-qi)
- [x] `nn.ConstantPad1d`
- [x] `nn.ConstantPad2d`
- [x] `nn.ConstantPad3d`
- [x] `nn.Conv1d` - Should be updated to support 2D input (@jbschlosser )
- [x] `nn.Conv2d` - Should be updated to support 3D input (@jbschlosser )
- [x] `nn.Conv3d` - Should be updated to support 4D input (@jbschlosser )
- [x] `nn.ConvTranspose1d` - Should be updated to support 2D input (@jbschlosser )
- [x] `nn.ConvTranspose2d` - Should be updated to support 3D input (@jbschlosser )
- [x] `nn.ConvTranspose3d` - Should be updated to support 4D input (@jbschlosser )
- [x] `nn.CosineEmbeddingLoss` - Should be updated to support 1D input with scalar target (@kshitij12345 )
- [x] `nn.CosineSimilarity` (pass `dim=0` for no-batch-dim support)
- [x] `nn.CrossEntropyLoss` - Should be updated to support 1D input with scalar target (@kshitij12345 )
- [x] `nn.Dropout`
- [x] `nn.Dropout2d`
- [x] `nn.Dropout3d`
- [x] `nn.ELU`
- [x] `nn.Embedding`
- [x] `nn.FeatureAlphaDropout` - Already supports no batch dims, but docs are missing
- [x] `nn.Flatten` (pass `start_dim=0` for no-batch-dim support)
- [x] `nn.Fold` - Should be updated to support 3D inputs (@kshitij12345)
- [x] `nn.FractionalMaxPool2d` - Should be updated to support 3D inputs
- [x] `nn.FractionalMaxPool3d` - Should be updated to support 4D inputs (@george-qi)
- [x] `nn.GELU`
- [x] `nn.GLU`
- [x] `nn.GRU` - Should be updated to support a single input sequence / hidden state; `batch_first` arg is meaningless in this case (@kshitij12345)
- [x] `nn.GRUCell` - Should be updated to support 1D input / hidden state (@kshitij12345)
- [x] `nn.GaussianNLLLoss` - Already supports no batch dims, but docs should make this clear (@george-qi)
- [x] `nn.Hardshrink`
- [x] `nn.Hardsigmoid`
- [x] `nn.Hardswish`
- [x] `nn.Hardtanh`
- [x] `nn.HingeEmbeddingLoss`
- [x] `nn.HuberLoss`
- [x] `nn.Identity`
- [x] `nn.InstanceNorm1d` - Should be updated to support 1D or 2D inputs (@kshitij12345)
- [x] `nn.InstanceNorm2d` - Should be updated to support 2D or 3D inputs (@kshitij12345)
- [x] `nn.InstanceNorm3d` - Should be updated to support 3D or 4D inputs (@kshitij12345)
- [x] `nn.KLDivLoss`
- [x] `nn.L1Loss`
- [x] `nn.LPPool1d` - Should be updated to support 2D inputs
- [x] `nn.LPPool2d`
- [x] `nn.LSTM` - Should be updated to support single input, hidden state, cell state; `batch_first` arg is meaningless in this case (@kshitij12345)
- [x] `nn.LSTMCell` - Should be updated to support 1D input, hidden state, cell state (@kshitij12345)
- [x] `nn.LayerNorm` (pass correctly-set `normalized_shape` for no-batch-dim support)
- [x] `nn.LazyConv1d` - Should be updated to support 2D input (@jbschlosser )
- [x] `nn.LazyConv2d` - Should be updated to support 3D input (@jbschlosser )
- [x] `nn.LazyConv3d` - Should be updated to support 4D input (@jbschlosser )
- [x] `nn.LazyConvTranspose1d` - Should be updated to support 2D input (@jbschlosser )
- [x] `nn.LazyConvTranspose2d` - Should be updated to support 3D input (@jbschlosser )
- [x] `nn.LazyConvTranspose3d` - Should be updated to support 4D input (@jbschlosser )
- [x] `nn.LazyLinear` - Already supports no batch dims, but docs should make this clear
- [x] `nn.LeakyReLU`
- [x] `nn.Linear` - Already supports no batch dims, but docs should make this clear
- [x] `nn.LogSigmoid`
- [x] `nn.LogSoftmax`
- [x] `nn.MSELoss` - Already supports no batch dims, but docs should make this clear
- [x] `nn.MarginRankingLoss` - Should be updated to support scalars (@kshitij12345)
- [x] `nn.MaxPool1d` - Already supports no batch dims when `return_indices=False`, but need to support `return_indices=True`
- [x] `nn.MaxPool2d` - Already supports no batch dims when `return_indices=False`, but need to support `return_indices=True`
- [x] `nn.MaxPool3d` - Already supports no batch dims when `return_indices=False`, but need to support `return_indices=True`
- [x] `nn.MaxUnpool1d` - Should be updated to support 2D input / indices
- [x] `nn.MaxUnpool2d` - Should be updated to support 3D input / indices
- [x] `nn.MaxUnpool3d` - Should be updated to support 4D input / indices
- [x] `nn.Mish`
- [x] `nn.MultiLabelMarginLoss`
- [x] `nn.MultiLabelSoftMarginLoss` - Should be updated to support 1D input / target
- [x] `nn.MultiMarginLoss` - Already supports no batch dims, but docs should make this clear
- [x] `nn.MultiheadAttention` - Should be updated to support unbatched query, key, values, and masks; `batch_first` arg is meaningless in this case
- [x] `nn.NLLLoss` - Should be updated to support 1D input and scalar target
- [x] `nn.PairwiseDistance` - Should be updated to support 1D inputs with scalar output
- [x] `nn.PixelShuffle`
- [x] `nn.PixelUnshuffle`
- [x] `nn.PoissonNLLLoss` - Already supports no batch dims, but docs should make this clear
- [x] `nn.RNN` - Should be updated to support a single input sequence / hidden state; `batch_first` arg is meaningless in this case (@kshitij12345)
- [x] `nn.RNNCell` - Should be updated to support 1D input and hidden state (@kshitij12345)
- [x] `nn.RReLU`
- [x] `nn.ReLU`
- [x] `nn.ReLU6`
- [x] `nn.ReflectionPad1d` - Should be updated to support 2D inputs
- [x] `nn.ReflectionPad2d`
- [x] `nn.ReflectionPad3d`
- [x] `nn.ReplicationPad1d` - Should be updated to support 2D inputs
- [x] `nn.ReplicationPad2d`
- [x] `nn.ReplicationPad3d`
- [x] `nn.SELU`
- [x] `nn.SiLU`
- [x] `nn.Sigmoid`
- [x] `nn.SmoothL1Loss` - Already supports no batch dims, but docs should make this clear
- [x] `nn.SoftMarginLoss`
- [x] `nn.Softmax`
- [x] `nn.Softmax2d` - Should be updated to support 3D input
- [x] `nn.Softmin`
- [x] `nn.Softplus`
- [x] `nn.Softshrink`
- [x] `nn.Softsign`
- [x] `nn.Tanh`
- [x] `nn.Tanhshrink`
- [x] `nn.Threshold`
- [x] `nn.Transformer` - Should be updated to support unbatched source / target sequences and masks (@kshitij12345)
- [x] `nn.TransformerDecoderLayer` - Should be updated to support unbatched inputs (@kshitij12345)
- [x] `nn.TransformerEncoderLayer` - Should be updated to support unbatched inputs (@kshitij12345)
- [x] `nn.TripletMarginLoss` - Should be updated to support 1D anchor, positive, and negative (@kshitij12345)
- [x] `nn.TripletMarginWithDistanceLoss` (support is based on the choice of distance function)
- [x] `nn.Unflatten` - Already supports no batch dims, but docs should make this clear
- [x] `nn.ZeroPad2d`

## Semantically incompatible modules
* `nn.BatchNorm1d` - Only defined over batch
* `nn.BatchNorm2d` - Only defined over batch
* `nn.BatchNorm3d` - Only defined over batch
* `nn.LazyBatchNorm1d` - Only defined over batch
* `nn.LazyBatchNorm2d` - Only defined over batch
* `nn.LazyBatchNorm3d` - Only defined over batch
* `nn.SyncBatchNorm` - Only defined over batch

## Modules that require BC-breaking changes
* `nn.ChannelShuffle` - While the docs say that the module accepts shape (*, C, H, W), the implementation assumes shape (N, C, *) with * being 1 or more dimensions; switching to (C, *) is BC-breaking because it would reinterpret dims
* `nn.EmbeddingBag` - Already supports 1D inputs with different semantics
* `nn.GroupNorm` - Supports arbitrary spatial dims; switching to (C, *) instead of (N, C, *) is BC-breaking since it would reinterpret dims
* `nn.LocalResponseNorm` - Supports (N, C, *) shape now; switching to (C, *) is BC-breaking since it would reinterpret dims
* `nn.PReLU` - Always assumes the 2nd dim is the channels dim
* `nn.Unfold` - Supports 4D inputs; switching to (C, *) instead of (N, C, *) is BC-breaking since it would reinterpret dims
* `nn.Upsample` - Supporting unbatched inputs would be BC-breaking because it would reinterpret dims

## Irrelevant modules
* `nn.Module`
* `nn.ModuleDict`
* `nn.ModuleList`
* `nn.ParameterDict`
* `nn.ParameterList`
* `nn.Sequential`
* `nn.TransformerDecoder` - see `nn.TransformerDecoderLayer` instead
* `nn.TransformerEncoder` - see `nn.TransformerEncoderLayer` instead

cc @albanD @mruberry @jbschlosser

### PR DESCRIPTION
**Pull request by thomasjpfan**
Fixes https://github.com/pytorch/pytorch/issues/60585

cc @albanD @mruberry @jbschlosser

### CODE REVIEW

File: torch/testing/_internal/common_nn.py [hunk 1]
```diff
@@ -2179,6 +2179,14 @@ def single_batch_reference_fn(input, parameters, module):
         input_fn=lambda: torch.rand(1, 3, 7, 7),
         desc='norm',
     ),
+    dict(
+        module_name='LPPool2d',
+        constructor_args=(1.5, 2),
+        cpp_constructor_args='torch::nn::LPPool2dOptions(1.5, 2)',
+        input_fn=lambda: torch.rand(3, 7, 7),
```
Comment (jbschlosser) [inline torch/testing/_internal/common_nn.py:2186]: nit: why `input_fn` here and `input_size` below?

**Review by jbschlosser — COMMENTED** [review]
Looking great! No changes needed for LPPool1d logic after all?

File: torch/testing/_internal/common_nn.py (hunk 1 above)
Comment (thomasjpfan) [reply torch/testing/_internal/common_nn.py]: Was trying to stay consistent with the `input_fn` above it. I updated the PR to just use `input_size`.

**Comment by thomasjpfan** [comment]
> No changes needed for LPPool1d logic after all?

`LPPool1d` support came for free when we updated average pool 1d. (`lppool1d` is defined in terms of `avg_pool1d`)

https://github.com/pytorch/pytorch/blob/acaac70f63da77d9145e20604b11262187d62d98/torch/nn/functional.py#L956

**Review by jbschlosser — APPROVED** [review]
LGTM!

**Comment by jbschlosser** [comment]
@thomasjpfan Test failures look real
